File: 2903-insert-greatest-common-divisors-in-linked-list/insert-greatest-common-divisors-in-linked-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function insertGreatestCommonDivisors(head: ListNode | null): ListNode | null {
    if (head == null || head.next == null) {
        return head
    }

    const gcd = (a, b) => {
        while (b !== 0) {
            [a, b] = [b, a % b]
        }
        return a
    }

    let prev = head
    let curr = head.next

    while (curr) {
        let num = gcd(prev.val, curr.val)
        let node = new ListNode(num)
        prev.next = node
        node.next = curr

        prev = curr
        curr = curr.next
    }

    return head
};

File: 2903-insert-greatest-common-divisors-in-linked-list/insert-greatest-common-divisors-in-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        prev = head
        curr = head.next

        while curr:
            n = math.gcd(prev.val, curr.val)
            prev.next = ListNode(n)
            prev.next.next = curr
            
            prev = prev.next.next
            curr = curr.next
        
        return head

        

    

    def hcd(self, a, b):
        if b == 0:
            return a
        else:
            return self.hcd(b, a % b)